### PR TITLE
Only admins can launch quick bar

### PR DESCRIPTION
**Pull request by donkawechico**
## Proposed change

Do not let non-admins launch quick bar.

Most of the items exposed in the quick-bar are not available to non-admin users:
1. An interactive list of all entities in the system
2. A list of reload commands

Ideally, the entity list would show all entities exposed to that user in their enabled dashboards, but I don't know of a way to do that right now.

As new commands get added to the command list, we can choose which to show to the user.

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by bramkragten** [comment]
Shouldnt we just disabled the quick bar for non admins, like not register the key?

**Comment by donkawechico** [comment]
> Shouldn't we just disabled the quick bar for non admins, like not register the key?

My thinking was that we would be soon letting non-admins use the navigation commands once that's merged (https://github.com/home-assistant/frontend/pull/7380).

That being said, permissions could get tricky _very_ quickly since action callbacks don't have any implicit sense of "permission" -- we just don't let users see the page containing those buttons.

I've convinced myself to just disable the bar entirely.

File: src/state/quick-bar-mixin.ts [hunk 1]
```diff
@@ -15,9 +15,36 @@ const isMacOS = /(Mac|iPhone|iPod|iPad)/i.test(navigator.platform);
...
 export default <T extends Constructor<HassElement>>(superClass: T) =>
   class extends superClass {
+    private isAdmin = false;
+
+    private haveCheckedAdminStatus = false;
+
+    protected updated(changedProps: PropertyValues) {
+      if (this.haveCheckedAdminStatus) {
+        return;
+      }
```
Comment (donkawechico) [inline src/state/quick-bar-mixin.ts:25]: I searched high and low for a reliable lifecycle hook that would *only* fire when the `this.hass.user` object was defined, but this is the only one that actually worked.

`firstUpdated` doesn't have user object at this point in bootstrapping.
`hassConnected` just tells us when hass object is available, but still doesn't guarantee a user object
`subscribeFrontendUserData()` seemed promising until I saw that it only fires when "show_advanced" is available.

So I just subscribed to all events and used a 2nd boolean to make sure that hass object updates don't do any extra work.

My worry is that with quick-bar-mixin being attached to such a high-level element, a lot of unnecessary checks will occur even though we've already done the work of attaching keyboard listeners.

Any advice?

File: src/state/quick-bar-mixin.ts [hunk 2]
```diff
@@ -15,9 +15,36 @@ const isMacOS = /(Mac|iPhone|iPod|iPad)/i.test(navigator.platform);
 
 export default <T extends Constructor<HassElement>>(superClass: T) =>
   class extends superClass {
+    private isAdmin = false;
+
+    private haveCheckedAdminStatus = false;
```
Comment (zsarnett) [inline src/state/quick-bar-mixin.ts:20]:
```suggestion
    private _isAdmin = false;

    private _haveCheckedAdminStatus = false;
```

File: src/state/quick-bar-mixin.ts [hunk 3]
```diff
@@ -15,9 +15,36 @@ const isMacOS = /(Mac|iPhone|iPod|iPad)/i.test(navigator.platform);
...
+
     protected firstUpdated(changedProps: PropertyValues) {
       super.firstUpdated(changedProps);
 
+      if (this.isAdmin) {
+        this.registerShortcut();
+      }
+    }
+
+    private registerShortcut() {
```
Comment (zsarnett) [inline src/state/quick-bar-mixin.ts:47]:
```suggestion
    private _registerShortcut() {
```

File: src/state/quick-bar-mixin.ts [hunk 4]
```diff
@@ -15,9 +15,36 @@ const isMacOS = /(Mac|iPhone|iPod|iPad)/i.test(navigator.platform);
 
 export default <T extends Constructor<HassElement>>(superClass: T) =>
   class extends superClass {
+    private _isAdmin = false;
+
+    private _haveCheckedAdminStatus = false;
+
+    protected updated(changedProps: PropertyValues) {
+      if (this.haveCheckedAdminStatus) {
```
Comment (zsarnett) [inline src/state/quick-bar-mixin.ts:23]:
```suggestion
      if (this._haveCheckedAdminStatus) {
```

File: src/state/quick-bar-mixin.ts [hunk 5]
```diff
@@ -15,9 +15,36 @@ const isMacOS = /(Mac|iPhone|iPod|iPad)/i.test(navigator.platform);
...
+            this.registerShortcut();
+          }
+        }
+      }
+    }
+
     protected firstUpdated(changedProps: PropertyValues) {
       super.firstUpdated(changedProps);
 
+      if (this.isAdmin) {
```
Comment (zsarnett) [inline src/state/quick-bar-mixin.ts:42]:
```suggestion
      if (this._isAdmin) {
```

File: src/state/quick-bar-mixin.ts [hunk 6]
```diff
@@ -15,9 +15,36 @@ const isMacOS = /(Mac|iPhone|iPod|iPad)/i.test(navigator.platform);
...
+        this.registerShortcut();
```
Comment (zsarnett) [inline src/state/quick-bar-mixin.ts:43]:
```suggestion
        this._registerShortcut();
```

File: src/state/quick-bar-mixin.ts (hunk 1 above)
Comment (zsarnett) [reply src/state/quick-bar-mixin.ts:25]: You are never un registering the shortcut if the user changes. Idk if that's possible without reconnecting and if its not then that's much easier.

Create a `shouldUpdate` return true only if `private _isRegestered = false` is false and if hass.user !== oldHass.user`

This way you arent updating all the time on every hass change

File: src/state/quick-bar-mixin.ts [hunk 7]
```diff
@@ -15,9 +15,36 @@ const isMacOS = /(Mac|iPhone|iPod|iPad)/i.test(navigator.platform);
...
       super.firstUpdated(changedProps);
 
+      if (this._isAdmin) {
+        this._registerShortcut();
+      }
+    }
+
+    private _registerShortcut() {
       document.addEventListener("keydown", (e: KeyboardEvent) => {
         if (this.isOSCtrlKey(e) && e.code === "KeyP") {
```
Comment (bramkragten) [inline src/state/quick-bar-mixin.ts:49]: If the timing is so hard, let's just add a `if (!this.hass?.user?.is_admin) { return; }` here.

File: src/state/quick-bar-mixin.ts [hunk 8]
```diff
@@ -15,9 +15,36 @@ const isMacOS = /(Mac|iPhone|iPod|iPad)/i.test(navigator.platform);
...
     protected firstUpdated(changedProps: PropertyValues) {
       super.firstUpdated(changedProps);
 
+      if (this._isAdmin) {
+        this._registerShortcut();
+      }
+    }
+
+    private _registerShortcut() {
       document.addEventListener("keydown", (e: KeyboardEvent) => {
```
Comment (bramkragten) [inline src/state/quick-bar-mixin.ts:48]: If the timing is so hard, let's just add a `if (!this.hass?.user?.is_admin) { return; }` here.

Comment (donkawechico) [reply src/state/quick-bar-mixin.ts:25]: Done

**Review by bramkragten — APPROVED** [review]

**Pull request merged**